### DISCUSSION TITLE
MacOS Keyboard bindings

### DISCUSSION THREAD

**Discussion by grbba** [daniele77/cli#99] [Q&A] [answered]
As i understood there is a history of commands so arrow up/down should recall commands in the stack. But on my Mac the up arrow e.g. produces
```
DccEX> color
Colors ON
DccEX> ^[[A
```
and nothing happens. Same for autocompletion using the tab key. The following is happening when in succession i type a tab and a back tab:
```
 - color
        Enable colors in the cli
 - nocolor
        Disable colors in the cli
 - build
        (menu)
 - cs
        (menu)
DccEX> co       ^[[Z
```
I would have expected autocompletion to give me color ?
//grbba

**Comment by daniele77** [comment, marked as answer]
Have you tried the `complete` or `pluginmanager` examples on MacOS?
Did they work as expected?

**Reply by grbba** [reply]
Yes i build all the examples and the `complete` works as expected no issues here. But in my app it goes wrong. I will compare my code with the complete one once again i may have missed something.
//

**Reply by daniele77** [reply]
Ok, that is important information: I don't have MacOS so I never tested personally the library on it. But I guess it should work being posix based.
So, the problem seems to be how you integrated the library in your application.
I just need a couple of additional information:
- which version of cli library are you trying (the head of the repository)?
- do you need remote session (i.e., telnet)?
- are you using boost asio or standalone asio libraries?

**Reply by grbba** [reply]
I cloned your repo a couple of days ago the id says faf07c5 which is the same as here on github in your repo for cli i.e. it should be the HEAD. 
I am using boost asio

Concerning telnet that may be an idea but i need a Telnet client then. I connect ( over serial or ethernet ) to the arduino. That will trigger some welcome message to be send and from then on i send commands to the Arduino and will receive back the answers - all text -. 
So far the idea was to spawn a 'reciever thread' before actually building the menus and have that wait/sleep as long as the port isn't open and wait sleep as long as there is no data.
on exit that thread, through a flag set to be true, will then exit and joined to exit the application. Maybe there is a better way from within your cli though and i am just doing things you can do already.

I just copied the way as its done in the 'complete' example ( i was just using the way of the simple example where the issue with the keys appeared ) but that didn't get ma any further as i get now the following at runtime:

`libc++abi.dylib: terminating with uncaught exception of type boost::wrapexcept<boost::system::system_error>: bind: Address already in use` 

as soon as i start the app but that may not have anything to do with cli

if you fancy to look at my code: [https://github.com/grbba/LayoutGraph](https://github.com/grbba/LayoutGraph)

It's in the `interactiveShell` branch `/src/DccShell.cpp`

**Reply by grbba** [reply]
Some more info 
once i commented out     
```
// cli::CliTelnetServer server(scheduler, 5000, cli);
// exit action for all the connections
// server.ExitAction( [](auto& out) { out << "Terminating this session...\n"; } );
```

the uncaught exception from above went away so yes then it dawned on that the 5000 is the port number for the server which is already in use... Once i changed it to a well known unused port it works :)

**Reply by daniele77** [reply]
Yes, the error you get a runtime was due to the TCP port already in use. You can change it (e.g., 5001) and the error should disappear.

**Reply by daniele77** [reply]
In your application, the special keys don't work just because you're using `CliFileSession`. You should use `CliLocalTerminalSession` instead (as in the examples `pluginmanager` and `complete`). 
`CliFileSession` is meant to be used when the input is feed from a file.  `CliLocalTerminalSession` is the right class to instantiate when you need to exploit all the features of OS shells.
Try with `CliLocalTerminalSession` and let me know.

Aside note: I see you're using asio asyncrhonous serial port. My advice is not to spawn another thread for the serial port, but to use the same of the boost scheduler you already have for the cli:

    boost::asio::io_context ioc;
    
    // use ioc for the serial port
    boost::asio::serial_port serial(ioc);
    ...

    // use the ioc for the cli scheduler
    BoostAsioScheduler scheduler(ioc);
    ...
    CliLocalTerminalSession localSession(cli, scheduler, std::cout, 200);
    ...
    ioc.run(); // use the current thread for both the serial port and for the cli

In this way, you're using the same thread for cli and serial port, so that you don't need synchronization between those parts.
(see `boostasio_nonowner_context.cpp` example).

In general, if you use `boost::asio` asynchronously in your application, you can have "concurrency" using just one thread by having one instance of `asio::io_context` that manages async events.

Another note: when your class has exclusive ownership of a pointer, use `std::unique_ptr` instead of raw pointer (e.g., in your pimpl code).

**Reply by grbba** [reply]
I managed to get it working once i really followed the complete example to the end ... I had also some issues using a thread started by a command from within the lamda of that command because of the use of boost:asio inside the function called when reading from a serial port. My assumption is that some resources never got freed in the serial port and despite being joinable the join blocked on exit somehow and when typing at that point an additional CR/LR I got the error from the library.

That has disappeared when i actually detached the thread ( which is possible as this thread only read incoming messages from the serial port and does nothing else )

**Reply by daniele77** [reply]
Ok, but please consider don't use a thread but the same boost::asio::io_context for both cli and serial port.

**Reply by grbba** [reply]
> In your application, the special keys don't work just because you're using `CliFileSession`. You should use `CliLocalTerminalSession` instead (as in the examples `pluginmanager` and `complete`).
> `CliFileSession` is meant to be used when the input is feed from a file. `CliLocalTerminalSession` is the right class to instantiate when you need to exploit all the features of OS shells.
> Try with `CliLocalTerminalSession` and let me know.
> 
> Aside note: I see you're using asio asyncrhonous serial port. My advice is not to spawn another thread for the serial port, but to use the same of the boost scheduler you already have for the cli:
> 
> ```
> boost::asio::io_context ioc;
> 
> // use ioc for the serial port
> boost::asio::serial_port serial(ioc);
> ...
> 
> // use the ioc for the cli scheduler
> BoostAsioScheduler scheduler(ioc);
> ...
> CliLocalTerminalSession localSession(cli, scheduler, std::cout, 200);
> ...
> ioc.run(); // use the current thread for both the serial port and for the cli
> ```
> 
> In this way, you're using the same thread for cli and serial port, so that you don't need synchronization between those parts.
> (see `boostasio_nonowner_context.cpp` example).
> 
> In general, if you use `boost::asio` asynchronously in your application, you can have "concurrency" using just one thread by having one instance of `asio::io_context` that manages async events.
> 
> Another note: when your class has exclusive ownership of a pointer, use `std::unique_ptr` instead of raw pointer (e.g., in your pimpl code).

Thanks a lot for this remark i will definitely try this out as it will make things simpler and more clear. I didn't see this reply before actually sending my comment of 2h ago. Somehow github hid it away from me and I discovered it in my mail :) 

Also the unique_ptr issue you raised i will implement. 

Again thanks for all your help 👍
